src/Features/Core/Portable/CodeRefactorings/IRefactoringHelpersService.cs

```diff
@@ -0,0 +1,95 @@
+        /// <para>
+        /// Note: this function strips all whitespace from both the beginning and the end of given <paramref name="selection"/>.
+        /// The stripped version is then used to determine relevant <see cref="SyntaxNode"/>. It also handles incomplete selections
+        /// of tokens gracefully.
+        /// </para>
+        /// </summary>
+        Task<TSyntaxNode> TryGetSelectedNodeAsync<TSyntaxNode>(Document document, TextSpan selection, CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode;
+
+        /// <summary>
+        /// <para>
+        /// Returns an instance of <typeparamref name="TSyntaxNode"/> for refactoring given specified selection in document or null
+        /// if no such instance exists.
+        /// </para>
+        /// <para>
+        /// A <typeparamref name="TSyntaxNode"/> instance is returned if:
+        /// - Selection is zero-width and inside/touching a Token with direct parent of type <typeparamref name="TSyntaxNode"/>.
+        /// - Selection is zero-width and touching a Token whose ancestor ends/starts precisely on current selection .
+        /// - Token whose direct parent of type <typeparamref name="TSyntaxNode"/> is selected.
+        /// - Whole node of a type <typeparamref name="TSyntaxNode"/> is selected.
+        /// </para>
+        /// <para>
+        /// The <paramref name="extractNode"/> function enables returning Nodes of type <typeparamref name="TSyntaxNode"/>
+        /// that are under those that might be selected / considered (as described above). It is a <see cref="Func{SyntaxNode, SyntaxNode}"/> that
+        /// should always return either given Node or a Node somewhere below it that should be tested for type <typeparamref name="TSyntaxNode"/> and
+        /// potentially returned instead of current Node. 
+        /// E.g. <see cref="ExtractNodeFromDeclarationAndAssignment{TNode}(SyntaxNode)"/>
+        /// allows returning right side Expresion node even if whole AssignmentNode is selected.
+        /// </para>
+        /// <para>
+        /// Note: this function strips all whitespace from both the beginning and the end of given <paramref name="selection"/>.
+        /// The stripped version is then used to determine relevant <see cref="SyntaxNode"/>. It also handles incomplete selections
+        /// of tokens gracefully.
+        /// </para>
+        /// </summary>
+        Task<TSyntaxNode> TryGetSelectedNodeAsync<TSyntaxNode>(Document document, TextSpan selection, Func<SyntaxNode, SyntaxNode> extractNode, CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode;
+
+        /// <summary>
+        /// Extractor function for <see cref="TryGetSelectedNodeAsync"/> methods that retrieves expressions from 
+        /// declarations and assignments. Otherwise returns unchanged <paramref name="node"/>.
+        /// </summary>
+        SyntaxNode ExtractNodeFromDeclarationAndAssignment<TNode>(SyntaxNode node) where TNode : SyntaxNode;
```

Comment:
I don't think this should be part of the interface.  It should be an inner impl detail.  Basically, inside the service we'll say "it's common to select this larger node while still intending to select this sub part.  so we check for that here blah blahblah"

Reply:
Noted, let's have this convo above.